### PR TITLE
Dts 26123

### PR DESCRIPTION
**Pull request by eswbogol**

### CODE REVIEW

File: database-scripts/upgrade-scripts/Release-8.0-from-7.0.js [hunk 1]
```diff
@@ -3688,4 +3688,53 @@ db.getCollection('kpi_category_mapping').insertOne( {
...
+ //DTS-26123 start
+ db.getCollection('field_mapping_structure').insertMany([
+ {
+        "fieldName" : "jiraDefectRejectionStatusKPI155",
+        	"fieldLabel" : "Ticket Rejected/Dropped Status",
+        	"fieldType" : "text",
+        	"fieldCategory" : "workflow",
+        	"section" : "WorkFlow Status Mapping",
+        	"tooltip" : {
+        		"definition" : "Status from workflow on which ticket is considered as Rejected/Dropped."
```
Comment (hirbabar) [inline database-scripts/upgrade-scripts/Release-8.0-from-7.0.js:3703]: missing kpi master data

File: database-scripts/archive/scripts/kpi_master_insert.js [hunk 2]
```diff
@@ -3007,6 +3007,30 @@ db.getCollection('kpi_master').insert(
...
+       "kpiId": "kpi155",
+       "kpiName": "Defect Count By Type",
+       "kpiUnit": "Count",
+       "isDeleted": "False",
+       "defaultOrder": 9,
+       "kpiCategory": "Backlog",
+       "kpiSource": "Jira",
+       "groupId": 11,
+       "thresholdValue": "",
+       "kanban": false,
```
Comment (hirbabar) [inline database-scripts/archive/scripts/kpi_master_insert.js:3021]: no do add in achive folder ....add in installtion script

File: database-scripts/archive/scripts/kpi_master_insert.js [hunk 3]
```diff
@@ -3007,6 +3007,30 @@ db.getCollection('kpi_master').insert(
...
+       "chartType": "pieChart",
+       "kpiInfo": {
+         "definition": "Total count of issues in the backlog with a breakup by defect type."
```
Comment (hirbabar) [inline database-scripts/archive/scripts/kpi_master_insert.js:3024]: add field mapping script for fresh installation script

File: database-scripts/upgrade-scripts/Release-8.0-from-7.0.js [hunk 4]
```diff
@@ -3688,4 +3688,85 @@ db.getCollection('kpi_category_mapping').insertOne( {
...
+ 			  "jiraDefectRejectionStatusKPI155": fm.jiraDefectRejectionStatus,
+ 			  "jiraDodKPI155": fm.jiraDod,
+ 			  "jiraLiveStatusKPI155": fm.jiraLiveStatus
+ 			  }
+         });
+ });
+
+const kpiIdsToCheck = ["kpi155"];
+var kpiData = db.getCollection('kpi_master').find( {kpiId: { $in: kpiIdsToCheck }}).toArray();
+if (kpiData.length === 0) {
```
Comment (hirbabar) [inline database-scripts/upgrade-scripts/Release-8.0-from-7.0.js:3743]: remove null check

**Review by pankumar8 — DISMISSED** [review]

**Review by hirbabar — DISMISSED** [review]

File: common/src/main/java/com/publicissapient/kpidashboard/common/model/application/dto/FieldMappingDTO.java [hunk 5]
```diff
@@ -375,6 +378,12 @@ public class FieldMappingDTO extends BasicModel {
 	private List<String> jiraIterationIssuetypeKPI120;
 	private List<String> jiraIterationIssuetypeKPI124;
 	private List<String> jiraIterationIssuetypeKPI39;
+	//DTS-26123 start
+	private String jiraDefectRejectionStatusKPI155;
+	@Builder.Default
+	private List<String> jiraDodKPI155 = Arrays.asList(DONE, CLOSED);
```
Comment (hirbabar) [inline common/src/main/java/com/publicissapient/kpidashboard/common/model/application/dto/FieldMappingDTO.java:384]: why pre defined list is added here ? can we add in metadata so running in processors is auto populate

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/util/KPIExcelUtility.java [hunk 6]
```diff
@@ -1524,4 +1524,39 @@ public static double roundingOff(double value) {
...
+				excelData.setStoryPoints(jiraIssue.getStoryPoints().toString());
+				List<String> sprintStatusList = Arrays.asList(CommonConstant.ACTIVE, CommonConstant.FUTURE);
+				excelData.setSprintName(StringUtils.isNotEmpty(jiraIssue.getSprintName())
+						&& StringUtils.isNotEmpty(jiraIssue.getSprintAssetState())
+						&& sprintStatusList.contains(jiraIssue.getSprintAssetState()) ? jiraIssue.getSprintName()
+								: "-");
+				String date = Constant.EMPTY_STRING;
+				if (jiraIssue.getCreatedDate() != null) {
+					date = DateUtil.dateTimeConverter(jiraIssue.getCreatedDate(), DATE_FORMAT_PRODUCTION_DEFECT_AGEING,
+							DateUtil.DISPLAY_DATE_FORMAT);
```
Comment (hirbabar) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/util/KPIExcelUtility.java:1549]: remove DATE_FORMAT_PRODUCTION_DEFECT_AGEING in this class and use generic DateUtill.DATE_FORMAT variable

File: database-scripts/installation-script/kpi-definitions.js [hunk 7]
```diff
@@ -3077,6 +3077,30 @@ db.getCollection('kpi_master').insertMany(
     "kpiFilter": "multiselectdropdown",
     "kpiWidth": 100,
     "calculateMaturity": false
+  },
+  {
+    "kpiId": "kpi155",
+    "kpiName": "Defect Count By Type",
+    "kpiUnit": "Count",
+    "isDeleted": "False",
+    "defaultOrder": 9,
```
Comment (hirbabar) [inline database-scripts/installation-script/kpi-definitions.js:3086]: verify is default Order 9 is okay or not

File: database-scripts/upgrade-scripts/Release-8.0-from-7.0.js [hunk 8]
```diff
@@ -3703,7 +3703,88 @@ db.getCollection('field_mapping_structure').insertMany([
...
+              $set: {
+ 			  "jiraDefectRejectionStatusKPI155": fm.jiraDefectRejectionStatus,
+ 			  "jiraDodKPI155": fm.jiraDod,
+ 			  "jiraLiveStatusKPI155": fm.jiraLiveStatus
+ 			  }
+         });
+ });
+
+const kpiIdsToCheck = ["kpi155"];
+var kpiData = db.getCollection('kpi_master').find( {kpiId: { $in: kpiIdsToCheck }}).toArray();
```
Comment (hirbabar) [inline database-scripts/upgrade-scripts/Release-8.0-from-7.0.js:3757]: remove null check from kpi master

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/DefectCountByTypeImpl.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+ *
+ */
+@Slf4j
+@Component
+public class DefectCountByTypeImpl extends JiraKPIService<Integer, List<Object>, Map<String, Object>> {
+
+	private static final String PROJECT_WISE_JIRA_ISSUE = "Jira Issue";
+
+	@Autowired
+	ConfigHelperService configHelperService;
```
Comment (hirbabar) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/DefectCountByTypeImpl.java:68]: use private

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/DefectCountByTypeImpl.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+			List<JiraIssue> jiraIssues = (List<JiraIssue>) resultMap.get(PROJECT_WISE_JIRA_ISSUE);
+			List<IterationKpiValue> filterDataList = new ArrayList<>();
+			Set<String> excludeStatuses = getExcludeStatuses(fieldMapping);
+			// method to exclude the issues with status
+			jiraIssues = getJiraIssueListAfterDefectsWithStatusExcluded(jiraIssues, excludeStatuses);
+			Map<String, List<JiraIssue>> statusWiseIssuesList = new HashMap<>(
+					CollectionUtils.isNotEmpty(jiraIssues)
+							? jiraIssues.stream().filter(
+									jiraIssue -> fieldMapping.getJiradefecttype().contains(jiraIssue.getOriginalType()))
+									.collect(Collectors.groupingBy(JiraIssue::getOriginalType))
```
Comment (hirbabar) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/DefectCountByTypeImpl.java:104]: if jiraissue.getOriginalType not existed for old data then check this map giving any errors?

File: customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/DefectCountByTypeImpl.java [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+	}
+
+	private static void getIssuesStatusCount(Map<String, List<JiraIssue>> statusData,
+			Map<String, Integer> statusWiseCountMap) {
+		for (Map.Entry<String, List<JiraIssue>> statusEntry : statusData.entrySet()) {
+			statusWiseCountMap.put(statusEntry.getKey(), statusEntry.getValue().size());
+		}
+	}
+
+	private void constructData(KpiElement kpiElement, String requestTrackerId, List<KPIExcelData> excelData,
```
Comment (hirbabar) [inline customapi/src/main/java/com/publicissapient/kpidashboard/apis/jira/scrum/service/DefectCountByTypeImpl.java:161]: add all methods java doc

**Review by hirbabar — APPROVED** [review]

**Review by gipathak — APPROVED** [review]

**Pull request merged**